### DISCUSSION TITLE
How to match PatchCore ONNX inference score with ckpt inference score?

### DISCUSSION THREAD

**Discussion by ricklina90** [open-edge-platform/anomalib#3644] [Q&A] [answered]
Hi Anomalib team,

First of all, thank you to the Anomalib team for providing such a powerful tool for industrial use.
I am trying to deploy PatchCore anomaly detection in a C# / LabVIEW application using ONNX runtime.
However, the inference scores from ONNX and ckpt are very different and I cannot match them.

## Environment
- anomalib: 2.5.0
- torch: 2.12.0+cu130
- onnxruntime:1.23.2
- OS: Windows

## Results Comparison (same image, same model)
```
ckpt : pred_score=0.6543, pred_label=True  (correct)
onnx : pred_score=1.0000, pred_label=True  (always clipped to 1.0)
```

## Pre-processor (from model.pre_processor)
```
PreProcessor(
  (transform): Compose(
    Resize(size=[256, 256], interpolation=InterpolationMode.BILINEAR, antialias=True)
    Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], inplace=False)
  )
  (export_transform): Compose(
    Resize(size=[256, 256], interpolation=InterpolationMode.BILINEAR, antialias=False)
    Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], inplace=False)
  )
)
```

## ONNX Output Nodes
```
['pred_score', 'pred_label', 'anomaly_map', 'pred_mask']
```

## Training Code
```python
datamodule = Folder(
    name="Epoch/Product1",
    root="./datasets/Epoch/Product1",
    normal_dir="train/good",
    normal_test_dir="test/good",
    abnormal_dir="test/defect",
    test_split_mode="from_dir",
    val_split_mode="same_as_test",
    train_batch_size=8,
    eval_batch_size=8,
)

model = Patchcore(
    backbone="wide_resnet101_2",
    num_neighbors=2,
    coreset_sampling_ratio=0.3,
)

engine = Engine(max_epochs=1)
engine.fit(datamodule=datamodule, model=model)
engine.export(
    model=model,
    export_type=ExportType.ONNX,
    ckpt_path="path/to/model.ckpt",
    export_root="path/to/output/",
)
```

## ckpt Inference Code (pred_score=0.6543, correct)
```python
model = Patchcore(backbone="wide_resnet101_2")
engine = Engine()
dataset = PredictDataset(path="test_image.png")

predictions = engine.predict(
    model=model,
    dataset=dataset,
    ckpt_path="path/to/model.ckpt",
)
print(predictions[0].pred_score)  # tensor([0.6543])
```

## ONNX Inference Code (pred_score=1.0, incorrect)
```python
import cv2
import numpy as np
import onnxruntime as ort
from PIL import Image

# Pre-processing
img_bgr = cv2.imread("test_image.png")
img_rgb = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2RGB)
pil_img = Image.fromarray(img_rgb)
pil_img = pil_img.resize((256, 256), Image.Resampling.BILINEAR)

img_float = np.array(pil_img).astype(np.float32) / 255.0
mean = np.array([0.485, 0.456, 0.406], dtype=np.float32)
std  = np.array([0.229, 0.224, 0.225], dtype=np.float32)
img_float = (img_float - mean) / std
img_transposed = np.transpose(img_float, (2, 0, 1))
input_tensor = np.expand_dims(img_transposed, axis=0)  # [1, 3, 256, 256]

sess = ort.InferenceSession("model.onnx")
outputs = sess.run(None, {sess.get_inputs()[0].name: input_tensor})
print("pred_score:", outputs[0])   # [1.0] ← incorrect
print("pred_label:", outputs[1])   # [True]
```

## What I have investigated
- ONNX initializer values are correct (match ckpt):
```
  post_processor._image_threshold : 0.9510
  onnx::Div_1516 (image_max-min)  : 2.9942
  onnx::Div_1519 (pixel_max-min)  : 6.6949
```
- Input tensors are nearly identical between ckpt and ONNX:
```
  ckpt : min=-2.1179, max=2.6400, mean=-1.3684
  onnx : min=-2.1179, max=2.6400, mean=-1.3684
```
- Since post_processor values and input are both correct,
  the raw score produced by the model backbone may be incorrect.

## Questions
1. Is the pre-processing code above correct for ONNX inference?
2. Is there any additional post-processing required after getting
   outputs from the ONNX model?
3. What is the correct way to verify the ONNX inference pipeline
   matches ckpt results?

**Comment by kirillmeisser** [comment]
Hi @ricklina90, I also have the same issue when exporting to ONNX. Have you found a solution?

**Comment by kirillmeisser** [comment, marked as answer]
I managed to get sameish outputs when comparing the output of the trained model and the ONNX export in the debugger. 

```python
dataset = PredictDataset(path=TEST_IMAGE_PATH)

predictions = engine.predict(
        model=model,
        dataset=dataset,
    )
```

and the resulting ONNX export using the following code

```python

def validate_onnx_export(onnx_path: Path, sample_image_path: Path, input_size: tuple[int, int] = (256, 256)):
    """Run ONNXRuntime inference on one sample image to validate exported model."""
    print(f"Validating ONNX model with ONNXRuntime: {onnx_path}")

    session = ort.InferenceSession(str(onnx_path), providers=["CPUExecutionProvider"])
    input_meta = session.get_inputs()[0]

    # Build input tensor from an RGB image as NCHW (1, 3, H, W).
    image = Image.open(sample_image_path).convert("RGB").resize(input_size)
    image_np = np.asarray(image)
    image_chw = np.transpose(image_np, (2, 0, 1))[None, ...]

    if input_meta.type == "tensor(float)":
        input_tensor = image_chw.astype(np.float32) / 255.0
    elif input_meta.type == "tensor(uint8)":
        input_tensor = image_chw.astype(np.uint8)
    else:
        raise ValueError(f"Unsupported ONNX input type: {input_meta.type}")

    outputs = session.run(None, {input_meta.name: input_tensor})

    print("ONNXRuntime inference successful.")
    print(f"  Input name: {input_meta.name}")
    print(f"  Input type: {input_meta.type}")
    print(f"  Input shape: {input_tensor.shape}")
    print(f"  Number of outputs: {len(outputs)}")
    for idx, out in enumerate(outputs):
        out_arr = np.asarray(out)
        print(
            f"    Output[{idx}] shape={out_arr.shape}, dtype={out_arr.dtype}, "
            f"min={out_arr.min():.6f}, max={out_arr.max():.6f}"
        )
    
    return outputs
```

I compared `predictions`and `outputs`. The values vary slightly but I believe that is caused by the ONNX export and should not affect the result meaningfully.

**Comment by ashwinvaidya17** [comment]
@ricklina90 thanks for the detailed reproducer and @kirillmeisser thanks for investigating.

@ricklina90 the root cause is double normalization.

The exported ONNX graph traces `AnomalibModule.forward`, which is:

```
pre_processor  →  model  →  post_processor
```

The `pre_processor` inside that graph performs **Resize + ImageNet mean/std normalization**. So the exported ONNX model already contains the normalize step as ONNX `Sub` / `Div` nodes.

Meanwhile, on the checkpoint path (`Engine.predict` / `predict_step`) the pre-processor is **not** called from `forward`. It runs once as a Lightning callback (`on_predict_batch_start`) before `self.model(batch.image)` is invoked.

Your inference script normalizes the tensor with the ImageNet stats **before** feeding it to ONNX Runtime. The baked-in `PreProcessor` then normalizes it **again**. After the second normalization the tensor lands roughly in `[-15, 10]` instead of `[-2.1, 2.6]`, all patch distances against the memory bank explode, and `PostProcessor._normalize` clamps everything to `1.0`, which is exactly the `pred_score=1.0000` and uniform `anomaly_map=1.0` you observed.

The reason your `min/max/mean` printouts matched on both sides is that you were comparing the tensor you pass into ORT with the tensor Lightning passes into `model.forward` — both are already normalized at that point. The doubled normalization happens *inside* the ONNX graph, after the input node.

The exported ONNX model expects a **raw `float32` tensor in `[0, 1]`, shape `(N, 3, H, W)`, RGB**. Remove the ImageNet mean/std subtraction from your inference script (also like @kirillmeisser mentioned):

```python
img = Image.open(img_path).convert("RGB")
x = np.array(img, dtype=np.float32) / 255.0              # keep [0, 1]
x = np.transpose(x, (2, 0, 1))[None, ...]                # NCHW
# do NOT subtract ImageNet mean / divide by std — the ONNX graph already does this
sess = ort.InferenceSession(str(onnx_path))
out = sess.run(None, {sess.get_inputs()[0].name: x})
```

**Comment by ricklina90** [comment]
Thank you all for the help and investigation!

Following the suggestion to remove the ImageNet mean/std normalization, 
I updated my inference code and got the following results:

**Results Comparison (same image, same model)**
| Method | pred_score | pred_label |
|---|---|---|
| ckpt | 0.6544 | True (Anomalous) ✅ |
| ONNX (with resize to 256x256) | 0.4760 | False (Normal) ❌ |
| ONNX (without resize) | 0.4897 | False (Normal) ❌ |

The pred_score is closer than before (previously always 1.0), 
but there is still a difference of ~0.18 compared to the ckpt result.
More importantly, the pred_label is incorrect — the ONNX model classifies 
the anomalous image as Normal, which would be unacceptable in a production environment.

**Test files for reproduction:**
🔗 https://1drv.ms/u/c/91da903f9f216365/IQA7vQCjRZ-lTI4RBSq1hXPGAfiqUWpI7rwSE3-u9-eEG7k

Download `anomalib_test.zip` and extract it.
The zip file is approximately 1.7 GB. We apologize for the large file size and appreciate your patience in downloading it.

**Folder structure after extraction:**
```
anomalib_test/
├── code/
│   └── testcode.py      # inference script for both ckpt and ONNX
├── image/
│   └── NG_x640.png      # test image (anomalous sample)
└── weights/
    ├── lightning/
    │   └── model.ckpt
    └── onnx/
        └── model.onnx
```

Note: Please update the paths in `testcode.py` to match your local environment.

**Environment:**
- anomalib : 2.5.0
- backbone : wide_resnet101_2
- OS       : Windows

I apologize if my testing is insufficient or if I am missing something.
Any further guidance would be greatly appreciated. Thank you!

**Comment by ricklina90** [comment]
Hi @ashwinvaidya17, Thank you for your previous explanation about double normalization.

I have followed your suggestion as described here:
🔗 https://github.com/open-edge-platform/anomalib/discussions/3644#discussioncomment-17664375

However, after removing the ImageNet mean/std normalization from my inference script, the pred_score still does not match the ckpt result:

- ckpt  : pred_score = 0.6544, pred_label = True  (Anomalous) ✅
- ONNX  : pred_score = 0.4760, pred_label = False (Normal)    ❌

I have uploaded test files (model.ckpt, model.onnx, test image, and inference script) to OneDrive for reproduction:
🔗 https://1drv.ms/u/c/91da903f9f216365/IQA7vQCjRZ-lTI4RBSq1hXPGAfiqUWpI7rwSE3-u9-eEG7k

Could you please advise what the correct inference pipeline should be?
Any further guidance would be greatly appreciated. Thank you!
